### PR TITLE
Building initrd

### PR DESCRIPTION
**Pull request by rafalcieslak**
This is my proposed solution for building the initrd with the kernel. The idea is to place some `initrd.rules` files in any place in the source tree. These files define what should be placed in the ramdisk - their format is self-explanatory, see e.g. `user/initrd.rules`:

```
prog/prog.uelf /bin/prog
fd_test/fd_test.uelf /bin/fd_test
test_fork/test_fork.uelf /bin/test_fork
misbehave/misbehave.uelf /bin/misbehave
```

The build script gathers these files and prepares the ramdisk according to these rules. It also generates a set of dependencies for `make`, so that it correctly updates the ramdisk when required.

This method is extensible, universal and simple to maintain.

@coodie, how do you like this script?

### CODE REVIEW

**Comment by coodie** [comment]
I don't really like using  Makefile (because It is old and hard to use), and I'd rather avoid using Makefile as much as possible. I'd rather have small bash script which builds initial ramdisk every time it should be built. In this case we have a problem where you want to load `.uelf` files which are inside our build system, therefore ramdisk should be rebuild every time they are changed. Although I hate Makefiles, I must admit that this solution is probably the cleanest, and your script is actually pretty readable.

**Comment by cahirwpz** [comment]
@coodie Please finish the review and approve changes if you don't have any more remarks.

File: update_initrd.sh [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+rm -rf $BUILDDIR
+mkdir -p $BUILDDIR
+
+SOURCEFILES=
+
+# Gather all files with rules, and copy files to the build dir
+for rulefile in $RULEFILES; do
+    echo "Processing file $rulefile"
+    sourcedir=$(dirname $rulefile)
+    rules=$(cat $rulefile)
```
Comment (cahirwpz) [inline update_initrd.sh]: You don't use this variable?

File: update_initrd.sh [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    while read -r line; do
+        source=$sourcedir/$(echo "$line" | cut  -d' ' -f1)
```
Comment (cahirwpz) [inline update_initrd.sh]: I prefer to use following syntax to access words within a line:

```
$ cat blah
foo bar boo
$ var=($(<blah)); echo "$var[1]:$var[2]:$var[3]"
foo:bar:boo
```

Splitting line into words is controlled by `IFS` shell variable (Internal Field Separator).

File: update_initrd.sh [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+DEPFILE=".initrd.D"
+CPIOFILE="initrd.cpio"
+
+### TODO: It is possible to implement this script so that it would only copy the
+### modified files, and incrementally update the ramdisk. However, this is
+### difficult, and we don't keep thousands of files in the ramdisk, so
+### recreating entire cpio archive will be good for now.
+
+# Clean/create the build directory
+rm -rf $BUILDDIR
```
Comment (cahirwpz) [inline update_initrd.sh]: Should not you clean up `BUILDDIR` after the job is done as well?

**Comment by rafalcieslak** [comment]
I've updated this branch to address mentioned problems, and to reflect on changes introduced by merging cpio support.

File: update_initrd.sh [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+### recreating entire cpio archive will be good for now.
+
+SOURCEFILES=
+
+# Gather all files with rules, and copy files to the build dir
+for rulefile in $RULEFILES; do
+    echo "Processing file $rulefile"
+    sourcedir=$(dirname $rulefile)
+    while read -r line; do
+        read source target <<< $line
```
Comment (cahirwpz) [inline update_initrd.sh]: Nice! I did not know this syntax!

**Comment by rafalcieslak** [comment]
So do we merge this?

**Pull request merged**